google-cloud-firestore/lib/google/cloud/firestore/document_reference.rb

```diff
@@ -0,0 +1,387 @@
+      #   # Get a document reference
+      #   nyc_ref = firestore.doc "cities/NYC"
+      #
+      class DocumentReference
+        ##
+        # @private The firestore client object.
+        attr_accessor :client
+
+        ##
+        # The document identifier for the document resource.
+        #
+        # @return [String] document identifier.
+        def document_id
+          path.split("/").last
+        end
+
+        ##
+        # A string representing the path of the document, relative to the
+        # document root of the database.
+        #
+        # @return [String] document path.
+        def document_path
+          path.split("/", 6).last
+        end
+
+        ##
+        # @private A string representing the full path of the document resource.
+        #
+        # @return [String] document resource path.
+        def path
+          @path
+        end
+
+        # @!group Access
+
+        ##
+        # Retrieves a list of collections nested under the document with data.
+        #
+        # @yield [collections] The block for accessing the collections.
+        # @yieldparam [CollectionReference] collection A collection.
+        #
+        # @return [Enumerator<CollectionReference>] collection list.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   nyc_ref.cols.each do |col|
+        #     puts col.collection_id
+        #   end
+        #
+        def cols
+          ensure_service!
+
+          return enum_for(:cols) unless block_given?
+
+          collection_ids = service.list_collections path
+          collection_ids.each { |collection_id| yield col(collection_id) }
+        end
+        alias_method :collections, :cols
+
+        ##
+        # Retrieves a collection nested under the document with data.
+        #
+        # @param [String] collection_path A string representing the path of
+        #   the collection, relative to the document.
+        #
+        # @return [CollectionReference] A collection.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a document reference
+        #   nyc_ref = firestore.doc "cities/NYC"
+        #
+        #   # Get precincts sub-collection
+        #   precincts_col = nyc_ref.col "precincts"
+        #
+        def col collection_path
+          if collection_path.to_s.split("/").count.even?
+            fail ArgumentError, "collection_path must refer to a collection."
+          end
+
+          CollectionReference.from_path "#{path}/#{collection_path}", client
+        end
+        alias_method :collection, :col
+
+        ##
+        # Retrieve the document data.
+        #
+        # @return [DocumentSnapshot] document with data.
```

Comment:
It might not have data if it doesn't exist.

Reply:
Changed to "document snapshot".